doc/specs/stdlib_hash_maps.md

```diff
@@ -0,0 +1,3088 @@
+
+## Licensing
+
+The Fortran Standard Library is distributed under the MIT License.
+However components of the library should be evaluated as to whether
+they are compatible with the MTI License.
+The current hash maps were inspired by an
+[implementation](http://chasewoerner.org/src/hasht/) of David
+Chase. While the code has been greatly modified from his
+implementation, he has give permission for the unrestricted use of
+his code.
+
+
+## The hash map modules
+
+The Fortran Standard Library provides three modules for the
+implementation of simple hash maps. These maps only accept hash
+functions with a single argument, the key, and that yield a 32 bit
+hash code, The modules will need to be modified to use hash functions
+with a different API. There are three modules:
+`stdlib_32_bit_key_data_wrapper`, `stdlib_chaining_hash_map` and
+`stdlib_open_hash_map`, corresponding to the files:
+`stdlib_32_bit_key_data_wrapper.f90`, `stdlib_chaining_hash_map.f90`,
+and `stdlib_open_hash_map.f90`. The module
+`stdlib_32_bit_key_data_wrapper` provides an interface to the 32 bit
+hash functions of the Standard Library module,
+`stdlib_32_bit_hash_functions`, and provides wrappers to some of the
+hash functions so that they no longer need to be supplied seeds. The
+module `stdlib_chaining_hash_map` defines a datatype,
+`chaining_hash_map_type`, implementing a simple separate chaining hash
+map noted more for its diagnostics than its performance. Finally the
+module, `stdlib_open_hash_map` defines a datatype,
+`open_hash_map_type`, implementing a simple open addressing hash
+map noted more for its diagnostics than its performance.
+
+These maps use separate chaining with linked lists and linear open
+addressing, respectively, to deal with hash index collisions, and are
+largely defined in the separated submodules, `stdlib_chaining_hash_maps`
+and `stdlib_open_hash_maps`, respectively.
+In `chaining_hash_map_type` the colliding indices are handled by using
+linked lists with their roots at the hash index.
+In `open_hash_map_type`, the colliding indices are handled by searching
+from the initial hash index in increasing
+steps of one (modulo the hash map size) for an open map bin.
+
+The maps share many attributes in common. The two types share a
+common Application Programers Interface (API). The maps use powers of
+two for their slot sizes, so that the function, `fibonacci_hash`, can
+be used to map the hash codes to indices in the map. This is
+expected to be more efficient than prime number mapping using a
+modulo operation, and reduces the requirement that the hash
+function need to do a good job randomizing its lower order bits.
+This requires a good randomizing hash method for good performance.
+Both adjust the map size to reduce collisions, based on 
+the ratio of the number of hash map probes to the number of subroutine 
+calls.
+The maps make extensive use of pointers internally, but a private
+finalization subroutine avoids memory leaks.
```

Comment:
Does this rely on the compiler to automatically call the finalization subroutine when a derived type goes out of scope? Last time I checked (albeit not recently) gfortran still didn't do this.

Reply:
Yes it relies on this to avoid memory leaks, though the largest amount of memory will be in arrays and not in the linked list used to avoid memory allocations. I don't think there are memory leaks with my most recent code compiled using the latest version of `gfortran`, though my testing was very indirect.